src/TurboRouter.sol

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.10;
+
+import {TurboMaster} from "./TurboMaster.sol";
+import {TurboSafe} from "./TurboSafe.sol";
+
+import {ENSReverseRecord} from "ERC4626/ens/ENSReverseRecord.sol";
+import {IERC4626, ERC4626RouterBase, IWETH9, PeripheryPayments} from "ERC4626/ERC4626RouterBase.sol";
+
+import {ERC20} from "solmate/tokens/ERC20.sol";
+import {SafeTransferLib} from "solmate/utils/SafeTransferLib.sol";
+
+import {ERC4626} from "solmate/mixins/ERC4626.sol";
+import {Auth, Authority} from "solmate/auth/Auth.sol";
+
+/**
+ @title a router which can perform multiple Turbo actions between Master and the Safes
+ @notice routes custom users flows between actions on the master and safes.
+         Accomodates use cases where the Safe and actions on the created Safe happen in the same transaction.
+         Supported Flows:
+          * creating a safe
+          * Auth methods (set owner/authority)
+          * ERC-4626 methods (deposit/withdraw/mint/redeem)
+          * TurboSafe methods (boost/less/sweep/slurp)
+ */
+contract TurboRouter is ERC4626RouterBase, ENSReverseRecord {
+    using SafeTransferLib for ERC20;
+
+    TurboMaster public immutable master;
+
+    constructor (TurboMaster _master, string memory name, IWETH9 weth) ENSReverseRecord(name) PeripheryPayments(weth) {
+        master = _master;
+    }
+
+    modifier authenticate(Auth target, bytes4 sig) {
+        Authority auth = target.authority();
+
+        require((address(auth) != address(0) && auth.canCall(msg.sender, address(target), sig)) || msg.sender == target.owner(), "not authed");
+
+        _;
+    }
+
+    function createSafeAndDeposit(ERC20 underlying, address to, uint256 amount, uint256 minSharesOut) external {
+        (TurboSafe safe, ) = master.createSafe(underlying);
+
+        safe.setOwner(msg.sender);
+
+        deposit(IERC4626(address(safe)), to, amount, minSharesOut);
+    }
+
+    function createSafeAndDepositAndBoost(
+        ERC20 underlying, 
+        address to, 
+        uint256 amount, 
+        uint256 minSharesOut, 
+        ERC4626 boostedVault, 
+        uint256 boostedFeiAmount
+    ) public {
+        (TurboSafe safe, ) = master.createSafe(underlying);
+
+        safe.setOwner(msg.sender);
+
+        deposit(IERC4626(address(safe)), to, amount, minSharesOut);
```

Comment:
should inline to avoid authenticate checks

Reply:
good point